server/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -1436,11 +1438,13 @@ public void processPendingDeletes(Index index, IndexSettings indexSettings, Time
                     }
                     if (remove.isEmpty() == false) {
                         logger.warn("{} still pending deletes present for shards {} - retrying", index, remove.toString());
-                        Thread.sleep(sleepTime);
+                        if (stopLatch.await(sleepTime, TimeUnit.MILLISECONDS)) {
+                            break;
+                        }
                         sleepTime = Math.min(maxSleepTimeMs, sleepTime * 2); // increase the sleep time gradually
                         logger.debug("{} schedule pending delete retry after {} ms", index, sleepTime);
                     }
-                } while ((System.nanoTime() - startTimeNS) < timeout.nanos());
+                } while ((System.nanoTime() - startTimeNS) < timeout.nanos() && lifecycle.started());
```

Comment:
Do we need this change now? Not really massively important, it just means that the logging is a little incorrect if we're stopped exactly between the `stopLatch.await()` timing out and getting to this check.